### PR TITLE
kie-issues#138: Create new feature on KIE Sandbox to allow the enabling/disabling of the PMML editors as a supported graphical editor

### LINKED ISSUE #138: Create new feature on KIE Sandbox to allow the enabling/disabling of the PMML editors as a supported graphical editor
Some distributions of KIE Sandbox might not be interested in allowing users to manually tweak PMML files on KIE Sandbox. We need a new environment variable to enable/disable the PMML Editor on KIE Sandbox. This env var must be usable on builds directly, or on `kie-sandbox-image`.

### PR DESCRIPTION
**Pull request by klbowers**
Closes https://github.com/kiegroup/kie-issues/issues/138

### CODE REVIEW

File: packages/kie-sandbox-image/README.md
Comment (tiagobento) [inline packages/kie-sandbox-image/README.md:1]: We need to add this new env var to the table here as well. Please follow the other variables and do a similar thing to this new one.

File: packages/online-editor/src/envelopeLocator/hooks/EditorEnvelopeLocatorContext.tsx [hunk 1]
```diff
@@ -38,3 +42,11 @@ export function EditorEnvelopeLocatorContextProvider(props: { children: React.Re
 export function useEditorEnvelopeLocator() {
   return useContext(EditorEnvelopeLocatorContext);
 }
+
+export function getEditorConfig() {
```
Comment (tiagobento) [inline packages/online-editor/src/envelopeLocator/hooks/EditorEnvelopeLocatorContext.tsx]: This is a custom hook, right? I mean, you're "abstracting" a `useEnv` and a `useMemo` here. I think this needs to be called `useEditorEnvelopeConfig`.

File: packages/online-editor/src/workspace/worker/sharedWorker.ts [hunk 2]
```diff
@@ -30,16 +31,24 @@ async function gitCorsProxyUrl(): Promise<string> {
...
-// eslint-disable-next-line prefer-const
 onconnect = async (e: MessageEvent) => {
-  console.log("Connected to Workspaces Shared Worker");
+  const editorEnvelopeConfigs = await fetchEditorEnvelopeConfig();
+  const editorEnvelopeLocator = new EditorEnvelopeLocatorFactory().create({
+    targetOrigin: "",
+    editorEnvelopeConfigs,
+  });
 
+  // eslint-disable-next-line prefer-const
```
Comment (tiagobento) [inline packages/online-editor/src/workspace/worker/sharedWorker.ts]: This line was before `onconnect = ...` before, right? We need to keep it in the same place.

File: packages/kie-sandbox-image/README.md [hunk 3]
```diff
@@ -237,3 +237,43 @@ Here's an example of what it should look like:
...
+    },
+    {
+      type: FileTypes.PMML,
+      filePathGlob: GLOB_PATTERN.pmml,
+      resourcesPathPrefix: "",
+      path: "pmml-envelope.html",
+      title: "Scorecard",
+      description: "PMML files are used to generate scorecards",
+    },
+```
```
Comment (tiagobento) [inline packages/kie-sandbox-image/README.md]: We're not in a JS context here. This needs to be plain JSON, so people configuring their env vars using a terminal will be able to use this one as reference.

Comment (tiagobento) [reply packages/kie-sandbox-image/README.md]: I.e., GLOB_PATTERN and FileTypes do not exist in your terminal.

File: packages/online-editor/src/envelopeLocator/EditorEnvelopeLocatorApi.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+export type EditorEnvelopeConfig = {
+  type: string;
+  filePathGlob: string;
+  resourcesPathPrefix: string;
+  path: string;
+  title: string;
+  description: string;
+};
```
Comment (tiagobento) [inline packages/online-editor/src/envelopeLocator/EditorEnvelopeLocatorApi.ts]:
```suggestion
export type EditorEnvelopeConfig = {
  extension: string;
  filePathGlob: string;
  editor: {
    resourcesPathPrefix: string;
    path: string;
  },
  card: {
    title: string;
    description: string;
  }
};
```

Please change this type to the structure above. It's easier to group related properties inside a nested object. Also change the necessary places after this change.

File: packages/online-editor/src/home/HomePage.tsx [hunk 5]
```diff
@@ -144,24 +146,21 @@ export function HomePage() {
...
-                  title={"Scorecard"}
-                  extension={"pmml"}
-                  description={"PMML files are used to generate scorecards"}
-                />
+                {editorConfig.map((config, index) => {
+                  return (
+                    <NewModelCard
+                      key={index}
+                      title={config.title}
+                      extension={config.type}
```
Comment (tiagobento) [inline packages/online-editor/src/home/HomePage.tsx]: This is a discrepancy in naming. `extension` receiving a `type`. See my suggestion below about changing the EditorEnvelopeConfig type.

**Review by tiagobento — CHANGES_REQUESTED** [review]
Thanks @kbowers-ibm ! Left some comments inline.

As a general comment, when introducing a new type or concept, we need to be careful to address it the same way everywhere. I see in this PR you're some times calling it `editorConfig`, and some times `editorEnvelopeConfig`. We need to choose one and make sure every place uses the same term.

More than that, maybe internally "Editor envelope config" makes sense to us, but for the outside (thinking about the env var here), we should not say "envelope". What do you think about simply `KIE_SANDBOX_EDITORS`? I mean, the fact that it is a env var already says it's a "config"... Then internally we can call it `EditorConfig`, following the same we did for `AuthProvider`.

Please make sure to change all the places with the new name we choose.

File: packages/online-editor/src/home/HomePage.tsx (hunk 5 above)
Comment (klbowers) [reply packages/online-editor/src/home/HomePage.tsx]: Fixed! The only thing now is that in the EditorEnvelopeLocatorFactory class we're now returning a new EnvelopeMapping with type: extension. Is it worth refactoring EnvelopeMapping from type to extension?
`return new EnvelopeMapping({
          type: config.extension,`

Comment (tiagobento) [reply packages/online-editor/src/home/HomePage.tsx]: If everywhere else was referring to this as `type`, then I think just the prop on `NewModelCard` that should be renamed to `type`, instead of renaming everything else to `extension`..

Comment (klbowers) [reply packages/online-editor/src/home/HomePage.tsx]: Explored renaming, but we decided that the amount of refactoring needed wasn't worth it

File: packages/kie-sandbox-image/README.md [hunk 6]
```diff
@@ -237,3 +239,58 @@ Here's an example of what it should look like:
...
+### Editors
+
+By default all three standard editors will be enabled (BPMN, DMN, PMML). To disable an editor simply delete/comment out the respective json.
+
+- **extension**: The extension of the file that you want to edit.
+- **filePathGlob**: The glob pattern of the file you want to edit.
+- **resourcesPathPrefix**: The path to the gwt-editor.
+- **path**: The path of the editor envelope.html.
+- **title**: The title of the editor that will be displayed on the home page.
+- **description**: Displays a short description of the editor under the title on the home page.
```
Comment (tiagobento) [inline packages/kie-sandbox-image/README.md]: This is not aligned with the recent changes introduced to the `EditorConfig` type.

File: packages/kie-sandbox-image/README.md [hunk 7]
```diff
@@ -54,14 +54,15 @@ This package contains the `Containerfile/Dockerfile` and scripts to build a cont
...
+   | :---------------------------------------------------------: | :---------------------------------------------------------------------------------------------------------------------------: | :-----------------------------------------------------------------: |
+   |             `KIE_SANDBOX_EXTENDED_SERVICES_URL`             |                                   The URL that points to the KIE Sandbox Extended Services.                                   | See [ defaultEnvJson.ts ](../online-editor/build/defaultEnvJson.ts) |
+   |              `KIE_SANDBOX_GIT_CORS_PROXY_URL`               |                         The URL that points to the Git CORS proxy for interacting with Git providers.                         | See [ defaultEnvJson.ts ](../online-editor/build/defaultEnvJson.ts) |
+   |        `KIE_SANDBOX_REQUIRE_CUSTOM_COMMIT_MESSAGES`         |                           Require users to type a custom commit message when creating a new commit.                           | See [ defaultEnvJson.ts ](../online-editor/build/defaultEnvJson.ts) |
```
Comment (tiagobento) [inline packages/kie-sandbox-image/README.md]: Can you please rename this to `KIE_SANDBOX_REQUIRE_CUSTOM_COMMIT_MESSAGE` without the `S` at the end? Me and Tim found this error this week.. Also add to the PR description that this unrelated change is present.

File: packages/online-editor/src/env/EnvJson.ts [hunk 8]
```diff
@@ -16,6 +16,7 @@
 
 import { AuthProvider } from "../authProviders/AuthProvidersApi";
 import { AcceleratorConfig } from "../accelerators/AcceleratorsApi";
+import { EditorConfig as EditorConfig } from "../envelopeLocator/EditorEnvelopeLocatorApi";
```
Comment (tiagobento) [inline packages/online-editor/src/env/EnvJson.ts]:
```suggestion
import { EditorConfig } from "../envelopeLocator/EditorEnvelopeLocatorApi";
```

That's VS Code's fault, I know :P

File: packages/online-editor/src/envelopeLocator/hooks/EditorEnvelopeLocatorContext.tsx [hunk 9]
```diff
@@ -18,16 +18,20 @@ import * as React from "react";
...
 
 export function EditorEnvelopeLocatorContextProvider(props: { children: React.ReactNode }) {
+  const editorConfig = useEditorConfig();
   const value = useMemo(
     () =>
       new EditorEnvelopeLocatorFactory().create({
         targetOrigin: window.location.origin,
+        editorConfig: editorConfig,
       }),
     []
```
Comment (tiagobento) [inline packages/online-editor/src/envelopeLocator/hooks/EditorEnvelopeLocatorContext.tsx]: Please always use the correct dependencies on hooks. Trust the linter! :)

You can also use a shorthand property here like 

```typescript
      new EditorEnvelopeLocatorFactory().create({
        targetOrigin: window.location.origin,
        editorConfig,
      })
```

File: packages/online-editor/src/envelopeLocator/hooks/EditorEnvelopeLocatorContext.tsx [hunk 10]
```diff
@@ -18,16 +18,20 @@ import * as React from "react";
...
 import { EditorEnvelopeLocatorFactory } from "../EditorEnvelopeLocatorFactory";
+import { useEnv } from "../../env/hooks/EnvContext";
+import { EditorConfig } from "../EditorEnvelopeLocatorApi";
 
 export type SupportedFileExtensions = "bpmn" | "bpmn2" | "BPMN" | "BPMN2" | "dmn" | "DMN" | "pmml" | "PMML";
 
 export const EditorEnvelopeLocatorContext = React.createContext<EditorEnvelopeLocator>({} as any);
 
 export function EditorEnvelopeLocatorContextProvider(props: { children: React.ReactNode }) {
+  const editorConfig = useEditorConfig();
```
Comment (tiagobento) [inline packages/online-editor/src/envelopeLocator/hooks/EditorEnvelopeLocatorContext.tsx]: Hmmm.. I think `editorsConfig` would be better, as this returns an array. Same for the name of the hook.

File: packages/online-editor/src/workspace/worker/sharedWorker.ts [hunk 11]
```diff
@@ -30,16 +31,24 @@ async function gitCorsProxyUrl(): Promise<string> {
...
+const workspaceServices = createWorkspaceServices({ gitCorsProxyUrl: gitCorsProxyUrl() });
 
 declare let onconnect: any;
 // eslint-disable-next-line prefer-const
 onconnect = async (e: MessageEvent) => {
-  console.log("Connected to Workspaces Shared Worker");
+  const editorConfig = await fetchEditorConfig();
+  const editorEnvelopeLocator = new EditorEnvelopeLocatorFactory().create({
+    targetOrigin: "",
+    editorConfig: editorConfig,
```
Comment (tiagobento) [inline packages/online-editor/src/workspace/worker/sharedWorker.ts]: Same here.

File: packages/online-editor/src/workspace/worker/sharedWorker.ts [hunk 12]
```diff
@@ -30,16 +31,24 @@ async function gitCorsProxyUrl(): Promise<string> {
   return env.KIE_SANDBOX_GIT_CORS_PROXY_URL;
 }
 
-const editorEnvelopeLocator = new EditorEnvelopeLocatorFactory().create({ targetOrigin: "" });
-const workspaceServices = createWorkspaceServices({ gitCorsProxyUrl: gitCorsProxyUrl() });
+async function fetchEditorConfig(): Promise<EditorConfig[]> {
```
Comment (tiagobento) [inline packages/online-editor/src/workspace/worker/sharedWorker.ts]:
```suggestion
async function fetchEditorsConfig(): Promise<EditorConfig[]> {
```

**Review by tiagobento — CHANGES_REQUESTED** [review]
Few minor issues commented inline. Ready to go when these are fixed and the PR is green. Thanks @kbowers-ibm !

**Review by tiagobento — APPROVED** [review]
